File: mobile-wallet-master/core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/client/FetchClientDetails.java
package org.mifos.mobilewallet.core.domain.usecase.client;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.entity.client.Client;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.utils.Constants;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 24/July/2018
 */

public class FetchClientDetails extends
        UseCase<FetchClientDetails.RequestValues, FetchClientDetails.ResponseValue> {

    private final FineractRepository mFineractRepository;

    @Inject
    public FetchClientDetails(FineractRepository fineractRepository) {
        mFineractRepository = fineractRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        mFineractRepository.getClientDetails(requestValues.clientId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<Client>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError(Constants.ERROR_FETCHING_CLIENT_DATA);
                    }

                    @Override
                    public void onNext(Client client) {
                        getUseCaseCallback().onSuccess(new ResponseValue(client));
                    }
                });

    }

    public static final class RequestValues implements UseCase.RequestValues {
        private final long clientId;

        public RequestValues(long clientId) {
            this.clientId = clientId;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private Client mClient;

        public ResponseValue(Client client) {
            mClient = client;
        }

        public Client getClient() {
            return mClient;
        }
    }
}

File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/client/FetchClientDetails.kt
package org.mifos.mobilewallet.core.domain.usecase.client

import org.mifos.mobilewallet.core.base.UseCase
import org.mifos.mobilewallet.core.data.fineract.entity.client.Client
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository
import org.mifos.mobilewallet.core.utils.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ankur on 24/July/2018
 */
class FetchClientDetails @Inject constructor(private val mFineractRepository: FineractRepository) :
    UseCase<FetchClientDetails.RequestValues, FetchClientDetails.ResponseValue>() {

    data class RequestValues(val clientId: Long) : UseCase.RequestValues
    data class ResponseValue(val client: Client) : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        mFineractRepository.getClientDetails(requestValues.clientId)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<Client>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(Constants.ERROR_FETCHING_CLIENT_DATA)
                }

                override fun onNext(client: Client) {
                    useCaseCallback.onSuccess(ResponseValue(client))
                }
            })
    }
}
